src/storage/txn/commands/mod.rs

```diff
@@ -432,3 +469,23 @@ impl Debug for Command {
         self.command_ext().fmt(f)
     }
 }
+
+pub trait ReadCommand<E: Engine>: CommandExt {
+    fn process_read(
+        &mut self,
+        snapshot: E::Snap,
+        statistics: &mut Statistics,
+    ) -> Result<ProcessResult>;
+}
+
+pub trait WriteCommand<S: Snapshot, L: LockManager, P: PdClient + 'static>: CommandExt {
```

Comment:
I think rather than introduce two new traits, it is better to add the process_* methods to the CommandExt trait and have them do nothing in the case that the command does not read or does not write. That might mean we have to return Option<Result<_>>, rather than just a Result.

Reply:
> I think rather than introduce two new traits, it is better to add the process_* methods to the CommandExt trait and have them do nothing in the case that the command does not read or does not write. That might mean we have to return Option<Result<_>>, rather than just a Result.

I wonder whether it is an error to eval `process_write` on a read-only command or vice versa.
Shall we panic or shall we return an `Option`?

Reply:
I think doing nothing and returning None is fine.

Reply:
> I think rather than introduce two new traits, it is better to add the process_* methods to the CommandExt trait and have them do nothing in the case that the command does not read or does not write. That might mean we have to return Option<Result<_>>, rather than just a Result.

@nrc
I still find merging these two traits into `CommandExt` nasty after some trys.
The major problem is about the type parameters, when I try adding `fn process_write<S: Snapshot, L: LockManager, P: PdClient + 'static>` into the `CommandExt` trait, the compiler give me [this error message](https://gist.github.com/longfangsong/7a627897a104334bcc8691c18910205a).
And then I tried to move the type parameters into the `CommandExt` trait, and I found I have to add these type parameters everywhere and [a compile error](https://gist.github.com/longfangsong/2bd73db81fd172d071339108c6c0c616) which I don't know how to fix still occurs.

Reply:
For the second error, I think you just need to add type args to`Command`, i.e., something like `... for Command<S, L, P>`. I think the first one is solved by putting the type args on the trait rather than on a method?

Reply:
> For the second error, I think you just need to add type args to`Command`, i.e., something like `... for Command<S, L, P>`. I think the first one is solved by putting the type args on the trait rather than on a method?

I'm afraid it needs to change too much code(might make almost every struct in `src/storage` need these type args, and affect many struct outside, eg. `LockManager` in `src/server`).
I doubt whether it worth the price.